config/base.py

```diff
@@ -0,0 +1,144 @@
+              entries. Ideally, you want to make it match the property's name
+              used when declaring the configuration implementation.
+    :arg name: a short, user friendly name, used to document the variable.
+    :arg description: a long description, used for documentation.
+    :arg values: optional list of possibly values. If specified, any other
+                 values set to the variable will result in an exception.
+    :arg constructor: optional constructor, used when accessing the variable.
+    :arg default: optional default value for this property.
+    """
+
+    def __init__(self, *, key: str, name: str, description: str = "",
+                 values: Optional[List[str]] = None,
+                 constructor: Optional[Callable[[str], T]] = None,
+                 default: Optional[str] = None):
+        self.key = key
+        self.name = name
+        self.description = description
+        self.values = values
+        self.constructor = constructor
+        self.default = default
+
+    def __set__(self, config: BaseConfig, value: str):
+        """Calls BaseConfig._set on the variable, saving it in the database."""
+        if self.values is not None and value not in self.values:
+            raise ValueError(
+                f"Value '{value}' is not matching the allowed "
+                f"values for '{self.key}'; expected one of"
+                f"[{', '.join(self.values)}]")
+        config._set(self.key, str(value))
+
+    def __get__(self, config: BaseConfig, objtype=None) -> Union[T, str]:
+        """Calls BaseConfig._get, retrieving it from the database."""
+        constructor = self.constructor or str
+        value = config._get(self.key) or self.default
+        if value is None:
+            raise KeyError(
+                f"No value assigned for the key {self.key} and no default "
+                "either.")
+        return constructor(value)
+
+
+class BaseConfig():
+    """Base class providing tooling to manage configuration.
+
+    Class inheriting this define configuration types (e.g. user config, server
+    config...) with a set of variables. Each instance of this configuration
+    represents the configuration for a specific entity (e.g. a specific user).
+
+    Child classes must:
+      - have a unique and defined CONFIGURATION_NAME, representing the table
+        storing this configuration.
+      - provide a unique ID for each entities they represent, representing a
+        row in the table.
+
+    Usage example:
+    >>> class UserConfig(BaseConfig):
+    ...     CONFIGURATION_NAME = "user"
+    ...
+    ...     prefered_meal = ConfigVariable(
+    ...         key = "prefered_meal",
+    ...         name = "Your prefered meal",
+    ...         description = "Whatever you like, we can make it, so say it!",
+    ...         values = ["pasta", "potatoes", "rice"],
+    ...         default = "pasta")
+    ...
+    >>> table = ConfigsTable(Database())
```

Comment:
(optional): missing `Database` import statement in example, not very important as the project is small, but I personally always find nice to clarify exactly from which Database class we are talking about.